Add documentation for additional events. (#3686)
boto/boto3: docs/source/guide/events.rst

| ``needs-retry``             | 7     | `Location <https://github.com/boto/botocore/blob/4691c4d6ac0cc3461fb5c40dc6908222866a525d/botocore/endpoint.py#LL353C1-L362C10>`_    |
+-----------------------------+-------+--------------------------------------------------------------------------------------------------------------------------------------+
| ``after-call``              | 8 *   | `Location <https://github.com/boto/botocore/blob/fc30d05149c248d5e601bea37be422e97c5ad7ee/botocore/client.py#LL947C1-L955C10>`_      |
+-----------------------------+-------+--------------------------------------------------------------------------------------------------------------------------------------+
| ``after-call-error``        | 8 *   | `Location <https://github.com/boto/botocore/blob/892c5bc73606ec02a6eff2522da1decff88d1c48/botocore/client.py#LL968C1-L975C14>`_      |
+-----------------------------+-------+--------------------------------------------------------------------------------------------------------------------------------------+

.. note::

  2 \* - ``creating-resource-class`` is emitted when using a service client and

>   ``provide-client-params`` is emitted when using a service resource.

Author: @dlm6693 These events pointed out here share the same order number because they're emitted in specific cased.  I'll try to make that more clear.

@nateprewitt  Thanks for catching! Thats was probably a mistake on my end when reordering the events in the notes section.

Boto3 emits a set of events that users can register to
customize clients or resources and modify the behavior of method calls.

Below is a table of events that users of Boto3 can register handlers to. More information
about each event can be found in the corresponding sections below:

+-----------------------------+-------+--------------------------------------------------------------------------------------------------------------------------------------+
| Event Name                  | Order | Emit Location                                                                                                                        |
+=============================+=======+======================================================================================================================================+

> | ``creating-client-class``   | 1     | `Location <https://github.com/boto/botocore/blob/aa63a377c6c9a78c87f55e5c0fc6ca6643198f3b/botocore/client.py#LL194C1-L198C10>`_      |

Author: I agree. I'll be replacing these links with ones of the following structure:

`https://github.com/search?q=repo%3A<github_repo>+<search_text>+path%3A<file_path_regex>&type=code`

Where:
> Note: All values are URL encoded.
* `<github_repo>` is the GitHub repo name such as `boto/botocore` or `boto/boto3`.
* `<search_text>` is the text/code we're searching for. This will probably be the event name. We may add more surrounding code if we need more context to narrow the search.
* `<file_path_regex>` is a regex for the path of the file we want to search for. Example: `/^botocore\/client\.py/`.

Example:
https://github.com/search?q=repo%3Aboto%2Fbotocore+creating-client-class+path%3A%2F%5Ebotocore%5C%2Fclient%5C.py%2F&type=code

> | ``creating-client-class``   | 1     | `Location <https://github.com/boto/botocore/blob/aa63a377c6c9a78c87f55e5c0fc6ca6643198f3b/botocore/client.py#LL194C1-L198C10>`_      |

Reviewer: Amazing thank you Jonathan!

Boto3 emits a set of events that users can register to
customize clients or resources and modify the behavior of method calls.

Here is a table of events that users of Boto3 can register handlers to. More information
about each event can be found in the corresponding sections below:

.. note::

  Events with a ``*`` in their order number are conditionally emitted while all others are always emitted.

>   An explanation of all 4 conditional events is provided below.

Reviewer: It looks like it's actually just 3 conditional events not 4

Author: yup, you're right

| ``needs-retry``             | 7     | `Location <https://github.com/search?q=repo%3Aboto%2Fbotocore+needs-retry+path%3A%2F%5Ebotocore%5C%2Fendpoint%5C.py%2F&type=code>`_           |
+-----------------------------+-------+-----------------------------------------------------------------------------------------------------------------------------------------------+
| ``after-call``              | 8 *   | `Location <https://github.com/search?q=repo%3Aboto%2Fbotocore+after-call.+path%3A%2F%5Ebotocore%5C%2Fclient%5C.py%2F&type=code>`_             |
+-----------------------------+-------+-----------------------------------------------------------------------------------------------------------------------------------------------+
| ``after-call-error``        | 8 *   | `Location <https://github.com/search?q=repo%3Aboto%2Fbotocore+after-call-error+path%3A%2F%5Ebotocore%5C%2Fclient%5C.py%2F&type=code>`_        |
+-----------------------------+-------+-----------------------------------------------------------------------------------------------------------------------------------------------+

.. note::

>   If any of the following keywords are included in an event's Full Event Name, you'll need

Author: no reason. I'll fix this.

:Full Event Name:
  ``'provide-client-params.service-name.operation-name'``

  Note: ``service-name`` refers to the value used to instantiate a client i.e.
  ``boto3.client('service-name')``. ``operation-name`` refers to the
  underlying API operation of the corresponding client method. To access
  the operation API name, retrieve the value from the
  ``client.meta.method_to_api_mapping`` dictionary using the name of the

>   desired client method as the key.

Author: Added this back in the latest commit since it already exists in the current docs and provides value to customers.

:Example:
  Here is an example of how to add a custom header before making an API call::

    import boto3

    s3 = boto3.client('s3')

    # Access the event system on the S3 client
    event_system = s3.meta.events

    # Create a function that adds a custom header and prints all headers.
    def add_custom_header_before_call(model, params, request_signer, **kwargs):
        params['headers']['my-custom-header'] = 'header-info'
        headers = params['headers']
        print(f'param headers: {headers}')

    #  Register the function to an event.
    event_system.register('before-call.s3.ListBuckets', add_custom_header_before_call)

    s3.list_buckets()

  This should output::

    param headers: { ... , 'my-custom-header': 'header-info'}

`request-created`
~~~~~~~~~~~~~~~~~~~~~

:Full Event Name:
  ``'request-created.service-name.operation-name'``

:Description:
  This event is emitted just after the request is created and triggers request signing.

:Keyword Arguments Emitted:

  :type request: ``botocore.awsrequest.AWSRequest``
  :param request: An AWSRequest object which represents the request that was
    created given some params and an operation model.

>   :type operation_name: str

Reviewer: Maybe we should add backticks to these like we do with other objects documented

    event_system.register('before-send.s3.ListBuckets', inspect_request_before_send)

    s3.list_buckets()

  This should output::

    request: <AWSPreparedRequest ... >

`needs-retry`
~~~~~~~~~~~~~~~~~~~~~~~

:Full Event Name:
  ``'needs-retry.service-name.operation-name'``

:Description:
  This event is emitted before checking if the most recent request needs to be retried.
  Use this event to define custom retry behavior when the configurable  
  `retry modes <https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html>`_ 
  are not sufficient.

:Keyword Arguments Emitted:

  :type response: tuple
  :param response: A tuple that includes both the ``botocore.awsrequest.AWSResponse``
    and a ``dict`` that represents the parsed response described by the model.

  :type endpoint: ``botocore.endpoint.Endpoint``
  :param endpoint:  Represents an endpoint for a particular service.

  :type operation: ``botocore.model.OperationModel``
  :param operation: A model representing the underlying API operation of the
    client method.

  :type attempts: int
  :param attempts: A number representing the amount of retries that have been attempted.

>   :type caught_exception: ``Exception`` | ``None```

Reviewer: ```suggestion
  :type caught_exception: ``Exception`` | ``None``
```

  :type model: ``botocore.model.OperationModel``
  :param model: A model representing the underlying API operation of the
    client method.

:Expected Return Value: ``None``

:Example:
  Here is an example of how to inject a parameter using the event::

    import boto3

    s3 = boto3.client('s3')

    # Access the event system on the S3 client
    event_system = s3.meta.events

    # Create a function
    def add_my_bucket(params, **kwargs):
        # Add the name of the bucket you want to default to.
        if 'Bucket' not in params:
            params['Bucket'] = 'mybucket'

    # Register the function to an event
    event_system.register('before-parameter-build.s3.ListObjectsV2', add_my_bucket)

    response = s3.list_objects_v2()

`before-call`
~~~~~~~~~~~~~~~~~~~~~

:Full Event Name:
  ``'before-call.service-name.operation-name'``

  .. note::

    ``service-name`` refers to the value used to instantiate a client i.e.
    ``boto3.client('service-name')``. ``operation-name`` refers to the
    underlying API operation name of the corresponding client method. 

:Description:
  This event is emitted just before creating the HTTP request and allows you to

>   modify request parameters prior to the request being created.

Reviewer: Ah yes this was a miss from my comment. I should know better considering the request compression project operates on `request_dict`/`params`!

And yeah to Jonas' point, we should call it something else to be a little clearer what data it actually contains. Maybe something along the lines of `A dictionary containing various components of an HTTP request.`

        print(f'operation_name: {operation_name}')

    session = Session()
    session.events.register('request-created.s3.ListObjectsV2', inspect_request_created)

    client = session.client('s3')
    client.list_objects_v2(Bucket='my-bucket')

  This should output::

    Request Info:
      method: GET
      url: https://my-bucket.s3...
      data: b'...'
      params: {...}
      auth_path: ...
      stream_output: ...
      headers: ...

    Operation Info:
      operation_name: ListObjectsV2

`before-send`
~~~~~~~~~~~~~~~~~~~~~

:Full Event Name:
  ``'before-send.service-name.operation-name'``

  .. note::

    ``service-name`` refers to the value used to instantiate a client i.e.
    ``boto3.client('service-name')``. ``operation-name`` refers to the
    underlying API operation name of the corresponding client method. 

:Description:
  This event is emitted when the operation has been fully serialized, signed,
  and is ready to be sent across the wire. This event allows the finalized
  request to be inspected and allows a response to be returned that fufills
  the request. If no response is returned botocore will fulfill the request

>   as normal.

Author: I believe the stubber uses `before-call` not `before-send`, but I did add it the the relevant section in the latest commit! Thanks!

:Keyword Arguments Emitted:

  :type request: ``botocore.awsrequest.AWSPreparedRequest``
  :param request: An object representing the properties of an HTTP request.

:Expected Return Value: ``None`` or an instance of :class:`.AWSResponse`

:Example:
  Here is an example of how to register a function that allows you to inspect
  the prepared request before it's sent::

    from boto3.session import Session

    def inspect_request_before_send(request, **kwargs):
        print(f'request:\n{request}')

    session = Session()
    session.events.register('before-send.s3.ListBuckets', inspect_request_before_send)

    client = session.client('s3')
    client.list_buckets()

  This should output::

    request:
    <AWSPreparedRequest ... >

`needs-retry`
~~~~~~~~~~~~~~~~~~~~~~~

:Full Event Name:
  ``'needs-retry.service-name.operation-name'``

  .. note::

    ``service-name`` refers to the value used to instantiate a client i.e.
    ``boto3.client('service-name')``. ``operation-name`` refers to the
    underlying API operation name of the corresponding client method. 

:Description:
  This event is emitted before checking if the most recent request needs to be retried.

>   Catching this event allows developers to customize retry behavior according to the
>   specific requirements of their application, rather than relying solely on Boto3's
>   default settings for retries.

Reviewer 2: ```suggestion
  Use this to define custom retry behavior when the configurable retry modes (https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html) are not sufficient.
```

:Keyword Arguments Emitted:

  :type response: tuple
  :param response: A tuple that includes both the ``botocore.awsrequest.AWSResponse``
    and a dict that represents the parsed version of the response.

  :type endpoint: ``botocore.endpoint.Endpoint``

>   :param endpoint:  Represents an endpoint for a particular service in a specific
>     region.

Reviewer: Some endpoints are global. Maybe it should just be:
```suggestion
  :param endpoint:  Represents an endpoint for a particular service.
```

  :type operation: ``botocore.model.OperationModel``
  :param operation: A model representing the underlying API operation of the
    client method.

  :type attempts: int
  :param attempts: A number representing the amount of retries that have been attempted.

  :type caught_exception: ``Exception``
  :param caught_exception: The exception raised after making an api call. If there was no

>     exception, this will be None.

Reviewer: Should the type be ``Exception`` | ``None`` then?

  :type request_dict: dict
  :param request_dict: A dictionary containing information related to the attempted API request.

:Expected Return Value: Return ``None`` if no retry is needed, or return an ``int`` representing the retry delay.

:Example:
  Here is an example of how to add a method to a resource class::

    from boto3.session import Session

    def needs_retry_handler(**kwargs):
        # Implement custom retry logic
        if some_condition:
            return None
        else:
            return some_delay

    session = Session()
    session.events.register('needs-retry', needs_retry_handler)

    client = session.client('s3')
    client.list_buckets()

`after-call`
~~~~~~~~~~~~~~~~~~~~~

:Full Event Name:
  ``'after-call.service-name.operation-name'``

  .. note::

    ``service-name`` refers to the value used to instantiate a client i.e.
    ``boto3.client('service-name')``. ``operation-name`` refers to the
    underlying API operation name of the corresponding client method. 

:Description:
  This event is emitted just after the service client makes an API call.

>   This event allows developers to post process or inspect the API response according to the

Reviewer 2: ```suggestion
  This event allows developers to postprocess or inspect the API response according to the
```

  specific requirements of their application if needed.

:Keyword Arguments Emitted:

  :type http_response: ``botocore.awsrequest.AWSResponse``
  :param http_response: A data class representing an HTTP response received from the server.

  :type parsed: dict
  :param params: A parsed version of the AWSResponse in the form of
    a python dictionary.

  :type model: ``botocore.model.OperationModel``
  :param model: A model representing the underlying API operation of the
    client method.

:Expected Return Value: ``None``

:Example:
  Here is an example that inspects args emitted from the ``after-call`` event::

    from boto3.session import Session

    def print_after_call_args(http_response, parsed, model, **kwargs):
        print(f'http_response:\n{http_response.text}')
        print(f'\nparsed:\n{parsed}')

>         print(f'\nmodel:\n{model.name}')

Reviewer: Separate print statements will already print on a new line so the extra newlines are unnecessary. It might make more sense to keep the key value pairs on the same line. At least for me that's a bit more readable.

Author: I simplified the example output format in the latest commit, thanks!

    session = Session()
    session.events.register('after-call.s3.ListObjectsV2', print_after_call_args)

    client = session.client('s3')
    client.list_objects_v2(Bucket='my-bucket')

  This should output::

>     http_response:
>     ...
> 
>     parsed:
>     ...

Reviewer: ```suggestion
    http_response:
    AWSResponse(...)

    parsed:
    {...}
```

    response = s3.list_objects()

`before-parameter-build`
~~~~~~~~~~~~~~~~~~~~~

:Full Event Name:
  ``'before-parameter-build.service-name.operation-name'``

  .. note::

    ``service-name`` refers to the value used to instantiate a client i.e.
    ``boto3.client('service-name')``. ``operation-name`` refers to the
    underlying API operation of the corresponding client method. To access
    the operation API name, retrieve the value from the
    ``client.meta.method_to_api_mapping`` dictionary using the name of the
    desired client method as the key.

:Description:
  This event is emitted before the API request parameters are built.

>   Use this event to inject or modify parameters after they're validated

Reviewer: Yeah Jonas is right. Input parameter validation happens during serialization, which is called in `BaseClient._convert_to_request_dict`. `before-parameter-build` is fired in `BaseClient._emit_api_params`, which is called earlier.

Could be worth mentioning this event is fired just before endpoint resolution, but not necessary.

  This event is emitted before checking if the most recent request needs to be retried.
  Catching this event allows developers to customize retry behavior according to the
  specific requirements of their application, rather than relying solely on Botocore's
  default settings for retrials.

:Keyword Arguments Emitted:

  :type response: tuple
  :param response: A tuple that includes both the ``botocore.awsrequest.AWSResponse``
    a dict that represents the parsed version of the response.

  :type endpoint: ``botocore.endpoint.Endpoint``
  :param endpoint:  Represents an endpoint for a particular service in a specific
    region.

  :type operation: ``botocore.model.OperationModel``
  :param operation: A model representing the underlying API operation of the
    client method.

  :type attempts: int
  :param attempts: A number representing the amount of retries that have been attempted.

  :type caught_exception: ``Exception``
  :param caught_exception: The exception raised after making an api call. If there was no
    exception, this will be None.

  :type request_dict: dict
  :param request_dict: A dictionary containing information related to the attempted API request.

> :Expected Return Value: Return ``None`` if no retry is needed, or return an int representing the retry delay.

Reviewer: ```suggestion
:Expected Return Value: Return ``None`` if no retry is needed, or return an ``int`` representing the retry delay.
```

`after-call`

> ~~~~~~~~~~~~

Reviewer: Can we just standardize how long these separators are? Everything else looks fine.

Reviewer: Per offline discussion it's difficult to standardize these because it must at least be as long as the length of the section header. If we chose an exact number it's always possible that we will have a section header that is longer than the number we choose. So it's probably best to make them the same length as the header.
